### PR TITLE
[WIP] docs: RHIDP-1474 Adding OAuth2 proxy as an authentication provider

### PR DESCRIPTION
**Pull request by themr0c**
docs: Adding OAuth2 proxy as an authentication provider

fixes https://issues.redhat.com/browse/RHIDP-1474

### CODE REVIEW

**Comment by themr0c** [comment]
Hi @kadel can you have a look at this? The current content is inspired from upstream. I expect the behavior to be different. Can you give me some hints?

File: modules/auth/proc-adding-oauth2-proxy-as-an-authentication-provider.adoc [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+The {product} `core-plugin-api` package includes an OAuth2-proxy authentication provider
+that can authenticate users
+by using an link:https://github.com/oauth2-proxy/oauth2-proxy[OAuth2 proxy] in front of an {product} instance.
+This enables to reuse existing authentications within a cluster.
+In general, the OAuth2 proxy supports all OpenID Connect providers.
+For more details, see link:https://oauth2-proxy.github.io/oauth2-proxy/configuration/providers/[the supported provider list].
+
+.Procedure
+
```
Comment (kadel) [inline modules/auth/proc-adding-oauth2-proxy-as-an-authentication-provider.adoc:12]: We should also document how to deploy oauth2-proxy as a sidecar container for RHDH

Comment (kadel) [reply modules/auth/proc-adding-oauth2-proxy-as-an-authentication-provider.adoc:12]: There is a blog post that documents how to do that: https://janus-idp.io/blog/2023/01/17/enabling-keycloak-authentication-in-backstage#deploying-backstage-using-the-backstage-helm-chart


TL;DR for helm chart it requries adding extra container to helm config

```yaml
  extraContainers:
    - name: oauth2-proxy
      env:
        - name: OAUTH2_PROXY_CLIENT_ID
          value: '{{ required "Keycloak Client Secret is Required" .Values.keycloak.clientId }}'
        - name: OAUTH2_PROXY_CLIENT_SECRET
          value: '{{ required "Keycloak Client Secret is Required" .Values.keycloak.clientSecret }}'
        - name: OAUTH2_PROXY_COOKIE_SECRET
          value: '{{ default (randAlpha 32 | lower | b64enc) .Values.keycloak.cookieSecret }}'
        - name: OAUTH2_PROXY_OIDC_ISSUER_URL
          value: '{{ required "Keycloak Issuer URL is Required" .Values.keycloak.issuerUrl }}'
        - name: OAUTH2_PROXY_SSL_INSECURE_SKIP_VERIFY
          value: 'true'
      ports:
        - name: oauth2-proxy
          containerPort: 4180
          protocol: TCP
      imagePullPolicy: IfNotPresent
      image: 'quay.io/oauth2-proxy/oauth2-proxy:latest'
      args:
        - '--provider=oidc'
        - '--email-domain=*'
        - '--upstream=http://localhost:7007'
        - '--http-address=0.0.0.0:4180'
        - '--skip-provider-button'
```

When installing via Operator this is currently not possible to do due to https://issues.redhat.com/browse/RHIDP-1760

File: modules/auth/proc-adding-oauth2-proxy-as-an-authentication-provider.adoc [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+. Set up the OAuth2-proxy authentication provider in the {product} backend (`app-config.yaml`).
++
+Add the `oauth2Proxy` provider configuration to your `app-config.yaml` under the root `auth` configuration:
++
+[source,yaml]
+----
+auth:
+  environment: development
```
Comment (kadel) [inline modules/auth/proc-adding-oauth2-proxy-as-an-authentication-provider.adoc:20]:
```suggestion
  environment: production
```

File: modules/auth/proc-adding-oauth2-proxy-as-an-authentication-provider.adoc [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+by using an link:https://github.com/oauth2-proxy/oauth2-proxy[OAuth2 proxy] in front of an {product} instance.
+This enables to reuse existing authentications within a cluster.
+In general, the OAuth2 proxy supports all OpenID Connect providers.
+For more details, see link:https://oauth2-proxy.github.io/oauth2-proxy/configuration/providers/[the supported provider list].
+
+.Procedure
+
+. Set up the OAuth2-proxy authentication provider in the {product} backend (`app-config.yaml`).
++
+Add the `oauth2Proxy` provider configuration to your `app-config.yaml` under the root `auth` configuration:
```
Comment (kadel) [inline modules/auth/proc-adding-oauth2-proxy-as-an-authentication-provider.adoc:15]: We should align working with other auth providers. For example in Azure auth provider ther is 

>Edit your custom {product-short} config map, such as `app-config-rhdh`.

>In the `app-config-rhdh.yaml` content, add the `microsoft` provider configuration >under the root `auth` configuration, and enable the `microsoft` provider for sign-in:

File: modules/auth/proc-adding-oauth2-proxy-as-an-authentication-provider.adoc [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+. To use the oauth2Proxy provider you must also configure it with a sign-in resolver.
++
+For the oauth2Proxy provider, the sign-in result is quite different than other providers.
+Because it's a proxy provider that can be configured to forward information through arbitrary headers, the auth result simply just gives you access to the HTTP headers of the incoming request.
+Using these you can either extract the information directly, or grab ID or access tokens to look up additional information and/or validate the request.
++
+A simple sign-in resolver might for example look like this:
++
+.`packages/backend/src/plugins/auth.ts`
```
Comment (kadel) [inline modules/auth/proc-adding-oauth2-proxy-as-an-authentication-provider.adoc:36]: In RHDH container it is not possible to edit files like this, there needs to be another way to configure this

File: modules/auth/proc-adding-oauth2-proxy-as-an-authentication-provider.adoc [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+To add the provider to the front-end, add the
+`ProxiedSignInPage` component to the sign-in page.
++
+To configure sign-in, provide a custom `SignInPage` app component.
+The app renders this component before any other routes.
+The component provides the identity of the current user.
+The `SignInPage` can render any number of pages and components, or just blank space with logic running in the background.
+In the end however it must provide a valid {product} user identity through the `onSignInSuccess` callback prop, at which point the rest of the app is rendered.
++
+The following example for OAuth2-proxy shows the additions needed to `packages/app/src/App.tsx`:
```
Comment (kadel) [inline modules/auth/proc-adding-oauth2-proxy-as-an-authentication-provider.adoc:68]: In RHDH container it is not possible to edit files like this, there needs to be another way to configure this

**Review by kadel — CHANGES_REQUESTED** [review]

**Comment by briandooley** [comment]
Triage: Closing as we do not currently support this.
